### PR TITLE
DLPX-85649 "test_oracle_smoke" failure on 5.15 kernel

### PR DESCRIPTION
**Pull request by prakashsurya**

### CODE REVIEW

**Comment by sebroy** [comment]
In terms of testing this change beyond the usual ab-pre-push tests, I would love to know if this changes/breaks the POSIX file locking (flock) semantics from the perspective of the NFS client(s). I'd love to do some manual tests using your configuration. E.g.:

From a single NFS client:
1. process A locks file X
2. process B locks file X (should fail)
3. process A releases the lock on file X
4. process B locks file X (should succeed)

Then, do the same thing with process A and B running on two different NFS clients. This one is more out of curiosity given that we don't typically have multiple clients grabbing locks on the same file.

File: files/common/lib/sysctl.d/50-override.conf [hunk 1]
```diff
@@ -93,3 +93,10 @@ fs.mount-max = 300000
 # value of 8192, resulting in upgrade failures.
 #
 fs.inotify.max_user_watches = 32768
+
+#
+# We've seen cases where an NFSv4 delegation is revoked by the server,
+# which then results in client IO errors. We disable NFSv4 delegation
+# entirely here, to avoid the problem.
+#
+fs.leases-enable = 0
```
Comment (ravi-cm) [inline files/common/lib/sysctl.d/50-override.conf:102]: Would this persist on a server reboot? I guess so but please confirm.

File: files/common/lib/sysctl.d/50-override.conf [hunk 2]
```diff
@@ -93,3 +93,10 @@ fs.mount-max = 300000
 # value of 8192, resulting in upgrade failures.
 #
 fs.inotify.max_user_watches = 32768
+
+#
+# We've seen cases where an NFSv4 delegation is revoked by the server,
```
Comment (ravi-cm) [inline files/common/lib/sysctl.d/50-override.conf:98]: I think, based on our tests, it was not clear if it's related to v 4.1 or just general v4 and above problem, right? If it is a client nfs version specific issue, wondering if we should just scope it 4.1 (which 6.10 uses) and leave it the default for other clients.  Not sure if we can even do that specific version based thing.
May be we should test on a platform other than Linux - Solaris x86 or HP just to make sure things are ok?

**Comment by prakashsurya** [comment]
Sure, let me do that testing..

Also, I kicked off some additional test runs:
```
> git-ab-pre-push --zfs-tests --zfs-perf-tests --io-stress
Detected linux-pkg package: delphix-platform
Your build is at: http://selfservice.jenkins.delphix.com/job/appliance-build-orchestrator-pre-push/5470/

> git-ab-pre-push --zfs-tests --zfs-perf-tests --io-stress --use-object-store -a
Detected linux-pkg package: delphix-platform
Your build is at: http://selfservice.jenkins.delphix.com/job/appliance-build-orchestrator-pre-push/5471/
```

File: files/common/lib/sysctl.d/50-override.conf (hunk 1 above)
Comment (prakashsurya) [reply files/common/lib/sysctl.d/50-override.conf:102]: Yes, that's the intention; I'll verify.

File: files/common/lib/sysctl.d/50-override.conf (hunk 2 above)
Comment (prakashsurya) [reply files/common/lib/sysctl.d/50-override.conf:98]: I don't know of a way to disable NFS delegations, outside of this tunable.. and this tunable is global.. so, while I wouldn't be opposed to set this on a specific NFS version, or specific client connect, I'm not sure it's possible to do that (currently, at least).

> May be we should test on a platform other than Linux - Solaris x86 or HP just to make sure things are ok?

Sure, how do I do this?

**Comment by prakashsurya** [comment]
### Test

From a single NFS client:

1. process A locks file X
2. process B locks file X (should fail)
3. process A releases the lock on file X
4. process B locks file X (should succeed)

### Experiment

- NFS mount setup on client:
```
$ mount | grep /mnt
ps-leases.dlpxdc.co:/public on /mnt type nfs (rw,vers=4,addr=10.110.220.152,clientaddr=10.110.200.81)
```

1. process A locks file X
```
$ time flock -n foo sleep 10

real    0m10.004s
user    0m0.001s
sys     0m0.001s

$ echo $?
0
```

2. process B locks file X (should fail)
```
$ time flock -n foo sleep 10

real    0m0.003s
user    0m0.000s
sys     0m0.001s

$ echo $?
1
```

4. process B locks file X (should succeed)
```
$ time flock -n foo sleep 10

real    0m10.005s
user    0m0.001s
sys     0m0.000s

$ echo $?
0
```

### Conclusion

This seems to be working correctly on a two process, single client configuration.

**Comment by prakashsurya** [comment]
> Then, do the same thing with process A and B running on two different NFS clients.

Here's results for the multi-client test...

- NFS mount set up on clients
```
[oracle@ip-10-110-200-81 mnt]$ mount | grep /mnt
ps-leases.dlpxdc.co:/public on /mnt type nfs (rw,vers=4,addr=10.110.220.152,clientaddr=10.110.200.81)
```
```
[oracle@ip-10-110-229-254 mnt]$ mount | grep /mnt
ps-leases.dlpxdc.co:/public on /mnt type nfs (rw,vers=4,addr=10.110.220.152,clientaddr=10.110.229.254)
```

1. process A locks file X
```
[oracle@ip-10-110-200-81 mnt]$ time flock -n foo sleep 10

real    0m10.006s
user    0m0.000s
sys     0m0.001s

[oracle@ip-10-110-200-81 mnt]$ echo $?
0
```

2. process B locks file X (should fail)
```
[oracle@ip-10-110-229-254 mnt]$ time flock -n foo sleep 10

real    0m0.011s
user    0m0.000s
sys     0m0.002s

[oracle@ip-10-110-229-254 mnt]$ echo $?
1
```

4. process B locks file X (should succeed)
```
[oracle@ip-10-110-229-254 mnt]$ time flock -n foo sleep 10

real    0m10.005s
user    0m0.001s
sys     0m0.000s

[oracle@ip-10-110-229-254 mnt]$ echo $?
0
```

Seems like this is working as well, using 2 processes, with each process on a different client.

File: files/common/lib/sysctl.d/50-override.conf (hunk 1 above)
Comment (prakashsurya) [reply files/common/lib/sysctl.d/50-override.conf:102]: Yes, it's persistent...
```
$ cat /proc/sys/fs/leases-enable
0

$ sudo reboot
Connection to ps-leases.dlpxdc.co closed by remote host.
Connection to ps-leases.dlpxdc.co closed.
⏎


> ssh delphix@ps-leases.dlpxdc.co
Warning: Permanently added 'ps-leases.dlpxdc.co' (ED25519) to the list of known hosts.

$ cat /proc/sys/fs/leases-enable
0

$ uptime
 20:19:38 up 3 min,  1 user,  load average: 2.57, 2.18, 0.96
 ```

File: files/common/lib/sysctl.d/50-override.conf (hunk 2 above)
Comment (ravi-cm) [reply files/common/lib/sysctl.d/50-override.conf:98]: From blackbox selfservice, 
- pick "oracle_mt_sanity" from the drop down for test scope and in the presented OS/platform configs, pick `ORACLE_19.16.0.0.0 | Solaris 11.4 | ASM | MT` as the test config.
- in addition, probably a good idea to run Oracle RAC as well. For this, from selfservice, pick "rac_mt_sanity" for test scope and pick `Oracle 12.2.0.1.0 | CENTOS 7.4 | CLONED | RAC(DBOMSRD914EA)` as the test config.

**Comment by prakashsurya** [comment]
> pick "oracle_mt_sanity" from the drop down for test scope and in the presented OS/platform configs, pick ORACLE_19.16.0.0.0 | Solaris 11.4 | ASM | MT as the test config.

See here: [Run 82483](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/82483/consoleFull), [Run 82782](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/82782/), [Run 83018](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/83018/), [Run 83661](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/83661/)

> in addition, probably a good idea to run Oracle RAC as well. For this, from selfservice, pick "rac_mt_sanity" for test scope and pick Oracle 12.2.0.1.0 | CENTOS 7.4 | CLONED | RAC(DBOMSRD914EA) as the test config.

See here: [Run 82484](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/82484/consoleFull), [Run 82806](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/82806/), [Run 83019](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/83019/), [Run 83663](http://selfservice.jenkins.delphix.com/job/blackbox-self-service/83663/)

**Review by sebroy — APPROVED** [review]

**Comment by prakashsurya** [comment]
@ravi-cm can you give those oracle runs a look? there's some test failures, but I'm not sure what a "good" run looks like.. should it be able to run without any failures?

**Comment by prakashsurya** [comment]
```
> git-ab-pre-push --zfs-tests --zfs-perf-tests --io-stress --use-object-store -a
Detected linux-pkg package: delphix-platform
Your build is at: http://selfservice.jenkins.delphix.com/job/appliance-build-orchestrator-pre-push/5476/
```
re-running this to see if it passes.. the first run, linked above, seems to have failed..

also, to help set a baseline, I've started a run with my changes reverted as well:
```
> git-ab-pre-push --zfs-tests --zfs-perf-tests --io-stress --use-object-store -a
Detected linux-pkg package: delphix-platform
Your build is at: http://selfservice.jenkins.delphix.com/job/appliance-build-orchestrator-pre-push/5477/
```

**Comment by prakashsurya** [comment]
ZFS performance tests the second time, so I think we're good w.r.t. performance.

**Comment by ravi-cm** [comment]
> @ravi-cm can you give those oracle runs a look? there's some test failures, but I'm not sure what a "good" run looks like.. should it be able to run without any failures?

I did not find any reason why we saw the timeout on those runs. Please re-run to see if it is consistent. The VM I revived is no longer there. So I will wait for the rerun links.

**Review by ravi-cm — APPROVED** [review]
ship-it once all automated runs are clean.

**Comment by prakashsurya** [comment]
> So I will wait for the rerun links.

I have 3 runs in my comment above.. I'll see if I can start a 4th for both of the two tests.

**Pull request merged**